File: lib/kmc/git_adapter.rb
module Kmc
  module GitAdapter
    class << self
      def init_repo(path)
        Dir.chdir(path) do
          `git init`
          `git config user.name KMC`
          `git config user.email kmc@kmc.kmc`
          `git config core.autocrlf false`

          File.open('.gitignore', 'w') do |file|
            file.write "!*\n"
          end
        end
      end

      def commit_everything(repo_path, commit_message)
        Dir.chdir(repo_path) do
          `git add -A -f`
          `git commit --allow-empty -m "#{commit_message}"`
        end
      end

      def revert_commit(repo_path, commit)
        Dir.chdir(repo_path) do
          # Favor "ours" (which is always HEAD for our purposes) when git-revert
          # can handle that on its own.
          `git revert --no-commit --strategy=merge --strategy-option=ours #{commit.sha}`

          # When files are being created or deleted, git will not do anything.
          # In this case, keep all files alive; better to accidentally pollute
          # than accidentally delete something important.
          `git add *`
        end
      end

      def list_commits(repo_path)
        Dir.chdir(repo_path) do
          `git log --oneline`.lines.map do |line|
            sha, message = line.split(' ', 2)
            Commit.new(message, sha)
          end
        end
      end

      def clone(repo_path, repo_url)
        Dir.chdir(repo_path) do
          `git clone #{repo_url} .`
        end
      end

      def pull(repo_path)
        Dir.chdir(repo_path) do
          `git pull`
        end
      end

      def changelog(repo_path, args = {})
        Dir.chdir(repo_path) do
          `git log --pretty=format:'%ai%  %s' -n10 #{args.join(" ")}`
        end
      end

    end

    class Commit < Struct.new(:message, :sha)
      def pre?
        type == :pre
      end

      def post?
        type == :post
      end

      def uninstall?
        type == :uninstall
      end

      def type
        # "POST: Example" --> :post
        message.split(':').first.downcase.to_sym
      end

      def subject
        # "POST: Example\n" --> "Example"
        message.split(' ', 2).last.strip
      end
    end
  end
end

File: lib/git_adapter.js
// Handles dealing with git.
var GitAdapter = {};

var Promise = require('bluebird');
var _ = require('lodash');
var fsPath = require('path');
var fs = Promise.promisifyAll(require('fs'));
var ChildProcess = Promise.promisifyAll(require('child_process'));

var Commit = require('./commit');

// Initializes a git repo at `path`.
GitAdapter.initRepo = function(path) {
  var cmd = [
    'git init',
    'git config user.name KMC',
    'git config user.email kmc@kmc.kmc',
    'git config core.autocrlf false'
  ].join(' ; ');

  var gitignorePath = fsPath.join(path, '.gitignore');

  return fs.writeFileAsync(gitignorePath, '!*\n').then(function() {
    return execInPath(path, cmd);
  });
}

// Adds all changes in `path` and commits them with the message `commitMessage`.
GitAdapter.commitEverything = function(path, commitMessage) {
  var cmd = [
    'git add -A -f',
    'git commit --allow-empty -m "' + commitMessage + '"'
  ].join(' ; ');

  return execInPath(path, cmd);
}

// Reverts a commit. This does not create a commit -- it merely changes the
// index to undo changes by a commit.
//
// The `commit` param must be a commit SHA.
GitAdapter.revertCommit = function(path, commit) {
  var cmd = [
    // Favor "our" (HEAD's) files.
    'git revert --no-commit --strategy=merge --strategy-option=ours ' + commit,

    // If there was a merge conflict or something, simply favor keeping files
    // around rather than deleting them.
    'git add *'
  ].join(' ; ');

  return execInPath(path, cmd);
}

// Returns a list of commits for a repo at `path`.
//
// The 'commits' returned are Commit objects.
GitAdapter.listCommits = function(path) {
  return execInPath(path, 'git log --oneline').then(function(output) {
    var stdout = output[0];
    var lines = stdout.split('\n');

    return _.chain(lines).filter(function(line) {
      return line.length > 0;
    }).map(function(line) {
      var splitAt = line.indexOf(' ');

      var sha = line.slice(0, splitAt);
      var message = line.slice(splitAt + 1);

      return new Commit(message, sha);
    }).value();
  });
}

// TODO: Add specs for #cloneRepo and #pullRepo where these actions can be
// mocked?

// Clones a git repo at `remoteUrl` into `localPath`.
GitAdapter.cloneRepo = function(localPath, remoteUrl) {
  return execInPath(localPath, 'git clone ' + remoteUrl + ' .');
}

// Pulls from a pre-configured git repo found locally at `path`.
GitAdapter.pullRepo = function(path) {
  return execInPath(path, 'git pull');
}

var execInPath = function(path, cmd) {
  var cmdInDir = 'cd ' + path + ' ; ' + cmd;

  return ChildProcess.execAsync(cmdInDir);
}

module.exports = GitAdapter;
